client/network-gossip/src/bridge.rs

```diff
@@ -178,6 +178,12 @@ impl<B: BlockT> Future for GossipEngine<B> {
 				ForwardingState::Idle => {
 					match this.network_event_stream.poll_next_unpin(cx) {
 						Poll::Ready(Some(event)) => match event {
+							Event::SyncConnected { remote } => {
```

Comment:
As far as I know `client/network-gossip` is only used in `client/finality-grandpa`. It is kept separate to potentially allow others to use it as well.

This change ties `client/network-gossip` directly to `client/finality-grandpa` as it enforces the users peer set to always equal the sync peer set.

I don't quite understand all changes in this pull request yet. In case we stick to this solution, would it not make sense to move `client/network-gossip` into `client/finality-grandpa` or at least call out that `client/network-gossip` is not supposed to be used by anyone other then `client/finality-grandpa`?

Reply:
Indeed, `network-gossip` now always maintains a set of peers equal to the sync one. The difference is that this is now decided by `network-gossip`, while before this was enforced by `sc-network`.

I don't think this means that `network-gossip` is tied to `finality-grandpa`, only that it is tied to the "main chain" the node manages.

I thought about this a bit and went for the solution of adding some documentation in https://github.com/paritytech/substrate/pull/7700/commits/9e702d6c99c948c7fe2f29e54db0da77981ccb49.